### PR TITLE
Implement Bag.selectUnique(). Fixes #506

### LINKED ISSUE #506: Implement Bag.selectUnique()
This should return a Set of all objects with count == 1 in the Bag.
Refer to #499 for details.

### PR DESCRIPTION
**Pull request by loradd**
Signed-off-by: Lorenzo Addazi <lorenzo.addazi@tutanota.com>

### CODE REVIEW

**Comment by nikhilnanivadekar** [comment]
@loradd Thanks for the PR. I see that you are returning a Bag. We would like to return a Set.
For that you need to add implementations in impl classes. For example in HashBag it will be something like this:

```
   @Override
    public MutableSet<T> selectUnique()
    {
        MutableSet<T> output = Sets.mutable.empty();
        bag.forEachWithOccurrences((each, occurrence) ->
        {
            if (occurrence == 1)
            {
                output.add(each);
            }
        });
        return output;
    }
```
For backwards compatibility the default implementation will just throw UnsupportedOperationException similar to: https://github.com/eclipse/eclipse-collections/blob/master/eclipse-collections-api/src/main/java/org/eclipse/collections/api/factory/list/MutableListFactory.java#L43

**Comment by loradd** [comment]
Okay @nikhilnanivadekar!

Just to be sure I understood all the changes involved.

> For backwards compatibility the default implementation will just throw UnsupportedOperationException similar to: https://github.com/eclipse/eclipse-collections/blob/master/eclipse-collections-api/src/main/java/org/eclipse/collections/api/factory/list/MutableListFactory.java#L43

Change org.eclipse.collections.api.bag.Bag#selectUnique() will throw UnsupportedOperationException.

> I see that you are returning a Bag. We would like to return a Set. For that you need to add implementations in impl classes.

Provide implementation returning Set instead of Bag in eclipse-collections and eclipse-collections-api.

Is it correct?

Thank you!

**Comment by nikhilnanivadekar** [comment]
Yes exactly!

**Comment by motlin** [comment]
To clarify, it’s SetIterable. That way we can do things like return sorted
or ordered sets from sorted or ordered bags. Thanks!

**Comment by loradd** [comment]
@motlin, thanks! What about mutability, is returning immutable or mutable sets depending on the filtered bag the approach you had in mind? Or should the result always be mutable?

**Comment by motlin** [comment]
Mutable or immutable depending on the type. Basically the same pattern as
select. If the return type of select varies from its super type, the return
type of selectUnique should too.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/ImmutableBag.java [hunk 1]
```diff
@@ -178,4 +178,7 @@
      */
     @Override
     ImmutableList<ObjectIntPair<T>> bottomOccurrences(int count);
+
+    @Override
+    ImmutableSet<T> selectUnique();
```
Comment (nikhilnanivadekar) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/ImmutableBag.java]: Needs default implementation which throws for backwards compatibility.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/ImmutableBagIterable.java [hunk 2]
```diff
@@ -71,4 +71,7 @@
 
     @Override
     MutableMapIterable<T, Integer> toMapOfItemToCount();
+
+    @Override
+    ImmutableSetIterable<T> selectUnique();
```
Comment (nikhilnanivadekar) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/ImmutableBagIterable.java]: Needs default implementation which throws for backwards compatibility.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/MutableBag.java [hunk 3]
```diff
@@ -188,4 +188,7 @@
      */
     @Override
     ImmutableBag<T> toImmutable();
+
+    @Override
+    MutableSet<T> selectUnique();
```
Comment (nikhilnanivadekar) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/MutableBag.java]: Needs default implementation which throws for backwards compatibility.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/MutableBagIterable.java [hunk 4]
```diff
@@ -116,4 +116,7 @@
 
     @Override
     MutableBagIterable<T> withoutAll(Iterable<? extends T> elements);
+
+    @Override
+    MutableSetIterable<T> selectUnique();
```
Comment (nikhilnanivadekar) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/MutableBagIterable.java]: Needs default implementation which throws for backwards compatibility.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/UnsortedBag.java [hunk 5]
```diff
@@ -128,4 +128,7 @@
 
     @Override
     <V> UnsortedBagMultimap<V, T> groupByEach(Function<? super T, ? extends Iterable<V>> function);
+
+    @Override
+    UnsortedSetIterable<T> selectUnique();
```
Comment (nikhilnanivadekar) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/UnsortedBag.java]: Needs default implementation which throws for backwards compatibility.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/sorted/ImmutableSortedBag.java [hunk 6]
```diff
@@ -77,6 +77,12 @@
         return this.selectByOccurrences(occurrences -> occurrences > 1);
     }
 
+    /**
+     * @since 9.2
+     */
+    @Override
+    ImmutableSortedSet<T> selectUnique();
```
Comment (nikhilnanivadekar) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/sorted/ImmutableSortedBag.java]: Needs default implementation which throws for backwards compatibility.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/sorted/MutableSortedBag.java [hunk 7]
```diff
@@ -63,6 +63,12 @@
         return this.selectByOccurrences(occurrences -> occurrences > 1);
     }
 
+    /**
+     * @since 9.2
+     */
+    @Override
+    MutableSortedSet<T> selectUnique();
```
Comment (nikhilnanivadekar) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/sorted/MutableSortedBag.java]: Needs default implementation which throws for backwards compatibility.

File: eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/sorted/SortedBag.java [hunk 8]
```diff
@@ -225,4 +225,7 @@
 
     @Override
     SortedBag<T> drop(int count);
+
+    @Override
+    SortedSetIterable<T> selectUnique();
```
Comment (nikhilnanivadekar) [inline eclipse-collections-api/src/main/java/org/eclipse/collections/api/bag/sorted/SortedBag.java]: Needs default implementation which throws for backwards compatibility.

File: eclipse-collections-code-generator/src/main/resources/api/bag/immutablePrimitiveBag.stg [hunk 9]
```diff
@@ -62,6 +63,9 @@ public interface Immutable<name>Bag extends Immutable<name>Collection, <name>Bag
         return this.selectByOccurrences(occurrences -> occurrences > 1);
     }
 
+    @Override
+    Immutable<name>Set selectUnique();
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/api/bag/immutablePrimitiveBag.stg]: Needs default implementation which throws for backwards compatibility.

File: eclipse-collections-code-generator/src/main/resources/api/bag/mutablePrimitiveBag.stg [hunk 10]
```diff
@@ -64,6 +65,9 @@ public interface Mutable<name>Bag extends Mutable<name>Collection, <name>Bag
         return this.selectByOccurrences(occurrences -> occurrences > 1);
     }
 
+    @Override
+    Mutable<name>Set selectUnique();
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/api/bag/mutablePrimitiveBag.stg]: Needs default implementation which throws for backwards compatibility.

File: eclipse-collections-code-generator/src/main/resources/api/bag/primitiveBag.stg [hunk 11]
```diff
@@ -65,6 +67,13 @@ public interface <name>Bag extends <name>Iterable
         return this.selectByOccurrences(occurrences -> occurrences > 1);
     }
 
+    /**
+     * Returns all elements of the bag that have exactly one occurrence.
+     *
+     * @since 9.2
+     */
+    <name>Set selectUnique();
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/api/bag/primitiveBag.stg]: Needs default implementation which throws for backwards compatibility.

File: eclipse-collections-code-generator/src/main/resources/impl/bag/immutable/immutablePrimitiveEmptyBag.stg [hunk 12]
```diff
@@ -182,6 +184,14 @@ final class Immutable<name>EmptyBag implements Immutable<name>Bag, Serializable
         return this;
     }
 
+    @Override
+    public Immutable<name>Set selectUnique()
+    {
+        Immutable<name>Bag selected = this.selectByOccurrences(occurrences -> occurrences == 1);
+        Immutable<name>Set result = <name>Sets.immutable.ofAll(selected);
+        return result;
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/impl/bag/immutable/immutablePrimitiveEmptyBag.stg]: Since this is an empty Bag, you can directly return `<name>Sets.immutable.empty()`

File: eclipse-collections-code-generator/src/main/resources/impl/bag/immutable/immutablePrimitiveHashBag.stg [hunk 13]
```diff
@@ -170,6 +172,14 @@ final class Immutable<name>HashBag implements Immutable<name>Bag, Serializable
         return this.delegate.selectByOccurrences(predicate).toImmutable();
     }
 
+    @Override
+    public Immutable<name>Set selectUnique()
+    {
+        Immutable<name>Bag selected = this.selectByOccurrences(occurrences -> occurrences == 1);
+        Immutable<name>Set result = <name>Sets.immutable.ofAll(selected);
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/impl/bag/immutable/immutablePrimitiveHashBag.stg]: This implementation is 2*O(n)
Efficient implementation will be:

```
Mutable<name>Set result = <name>Sets.mutable.empty();
this.forEachWithOccurrences((each, occurrences) -> 
    {
        if(occurrences == 1)
        {
            result.add(each);
        }
    });
return result.toImmutable();
```

File: eclipse-collections-code-generator/src/main/resources/impl/bag/immutable/immutablePrimitiveSingletonBag.stg [hunk 14]
```diff
@@ -169,6 +171,14 @@ final class Immutable<name>SingletonBag implements Immutable<name>Bag, Serializa
             : <name>Bags.immutable.empty();
     }
 
+    @Override
+    public Immutable<name>Set selectUnique()
+    {
+        Immutable<name>Bag selected = this.selectByOccurrences(occurrences -> occurrences == 1);
+        Immutable<name>Set result = <name>Sets.immutable.ofAll(selected);
+        return result;
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/impl/bag/immutable/immutablePrimitiveSingletonBag.stg]: This is a SingletonBag, so it will contain only 1 element and occurrences == 1
So, you can do this `return <name>Sets.immutable.with(this.value);`

File: eclipse-collections-code-generator/src/main/resources/impl/bag/mutable/primitiveHashBag.stg [hunk 15]
```diff
@@ -224,6 +227,14 @@ public class <name>HashBag
         return result;
     }
 
+    @Override
+    public Mutable<name>Set selectUnique()
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/impl/bag/mutable/primitiveHashBag.stg]: This is 2*O(n)
Efficient implementation will be:
```

Mutable<name>Set result = <name>Sets.mutable.empty();
this.forEachWithOccurrences((each, occurrences) -> 
    {
        if(occurrences == 1)
        {
            result.add(each);
        }
    });
return result;
```

File: eclipse-collections-code-generator/src/main/resources/impl/bag/mutable/synchronizedPrimitiveBag.stg [hunk 16]
```diff
@@ -169,6 +171,14 @@ public class Synchronized<name>Bag
         }
     }
 
+    @Override
+    public Mutable<name>Set selectUnique()
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/impl/bag/mutable/synchronizedPrimitiveBag.stg]: Since this is synchronized, you need to synchronize it on the lock. Also, you can delegate it to the underlying Bag.

```
synchronized (this.getLock())
{
    Mutable<name>Set result = this.getMutable<name>Bag().selectUnique();
}
return result;
```

File: eclipse-collections-code-generator/src/main/resources/impl/bag/mutable/unmodifiablePrimitiveBag.stg [hunk 17]
```diff
@@ -110,6 +112,14 @@ public class Unmodifiable<name>Bag
         return this.getMutable<name>Bag().selectByOccurrences(predicate);
     }
 
+    @Override
+    public Mutable<name>Set selectUnique()
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/impl/bag/mutable/unmodifiablePrimitiveBag.stg]: Similar to Synchronized, you can delegate it to the underlying MutableBag, it changes it to one line implementation.

`return this.getMutable<name>Bag().selectUnique();`

File: eclipse-collections-code-generator/src/main/resources/test/bag/immutable/abstractImmutablePrimitiveBagTestCase.stg [hunk 18]
```diff
@@ -89,6 +90,13 @@ public abstract class AbstractImmutable<name>BagTestCase extends AbstractImmutab
         Assert.assertEquals(<name>HashBag.newBagWith(<["2", "2", "3", "3", "3"]:(literal.(type))(); separator=", ">), bag.selectDuplicates());
     }
 
+    @Test
+    public void selectUnique()
+    {
+        Immutable<name>Bag bag = this.newWith(<["1", "2", "2", "3", "3", "3"]:(literal.(type))(); separator=", ">);
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/test/bag/immutable/abstractImmutablePrimitiveBagTestCase.stg]: Please add this also as a test case

```
Immutable<name>Bag bag2 = this.newWith(<["1", "2", "2", "3", "3", "3", "4", "5", "5", "6"]:(literal.(type))(); separator=", ">);
Assert.assertEquals(<name>Sets.immutable.of(<["1", "4", "6"]:(literal.(type))(); separator=", ">), bag2.selectUnique());
```

File: eclipse-collections-code-generator/src/main/resources/test/bag/mutable/abstractMutablePrimitiveBagTestCase.stg [hunk 19]
```diff
@@ -83,6 +84,13 @@ public abstract class AbstractMutable<name>BagTestCase extends AbstractMutable<n
         Assert.assertEquals(<name>HashBag.newBagWith(<["2", "2", "3", "3", "3"]:(literal.(type))(); separator=", ">), bag.selectDuplicates());
     }
 
+    @Test
+    public void selectUnique()
+    {
+        Mutable<name>Bag bag = this.newWith(<["1", "2", "2", "3", "3", "3"]:(literal.(type))(); separator=", ">);
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/test/bag/mutable/abstractMutablePrimitiveBagTestCase.stg]: Similar to abstractImmutablePrimitiveBagTestCase, can you please add this test case to mutable side:

```
Mutable<name>Bag bag2 = this.newWith(<["1", "2", "2", "3", "3", "3", "4", "5", "5", "6"]:(literal.(type))(); separator=", ">);
Assert.assertEquals(<name>Sets.mutable.of(<["1", "4", "6"]:(literal.(type))(); separator=", ">), bag2.selectUnique());
```

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/AbstractBag.java [hunk 20]
```diff
@@ -541,7 +541,7 @@ public long sumOfLong(LongFunction<? super T> function)
         long[] sum = {0L};
         this.forEachWithOccurrences((each, occurrences) -> {
             long longValue = function.longValueOf(each);
-            sum[0] += longValue * (long) occurrences;
+            sum[0] += longValue * occurrences;
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/AbstractBag.java]: This change seems to be unrelated to this PR. Can you please revert it? Or it was due to some checkstyle violation?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/immutable/AbstractImmutableBag.java [hunk 21]
```diff
@@ -240,4 +242,12 @@ public ImmutableShortBag collectShort(ShortFunction<? super T> shortFunction)
         }
         return result.toImmutable();
     }
+
+    @Override
+    public ImmutableSet<T> selectUnique()
+    {
+        ImmutableBag<T> selected = this.selectByOccurrences(occurrences -> occurrences == 1);
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/immutable/AbstractImmutableBag.java]: Similar to review comments of primitive side, this can be made efficient by using forEachWithOccurrences().

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/AbstractMutableBag.java [hunk 22]
```diff
@@ -274,4 +275,12 @@ public MutableShortBag collectShort(ShortFunction<? super T> shortFunction)
         }
         return new NonParallelUnsortedBag<>(this);
     }
+
+    @Override
+    public MutableSet<T> selectUnique()
+    {
+        MutableBag<T> selected = this.selectByOccurrences(occurrences -> occurrences == 1);
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/AbstractMutableBag.java]: This can be made efficient by using forEachWithOccurrences() similar to the primitive side comments.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/MultiReaderHashBag.java [hunk 23]
```diff
@@ -1201,6 +1202,14 @@ public String toStringOfItemToCount()
         {
             return (MutableBag<T>) this.delegate;
         }
+
+        @Override
+        public MutableSet<T> selectUnique()
+        {
+            MutableBag<T> selected = this.selectByOccurrences(occurrences -> occurrences == 1);
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/MultiReaderHashBag.java]: This can be delegated to underlying Bag by calling 
```
        this.acquireReadLock();
        try
        {
            return this.delegate.selectUnique();
        }
        finally
        {
            this.unlockReadLock();
        }
```

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/MultiReaderHashBag.java [hunk 24]
```diff
@@ -1236,4 +1245,12 @@ public void becomeUseless()
             this.delegate = null;
         }
     }
+
+    @Override
+    public MutableSet<T> selectUnique()
+    {
+        MutableBag<T> selected = this.selectByOccurrences(occurrences -> occurrences == 1);
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/MultiReaderHashBag.java]: This can be delegated:
`return this.getDelegate().selectUnique();`

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/SynchronizedBag.java [hunk 25]
```diff
@@ -425,4 +426,12 @@ public MutableShortBag collectShort(ShortFunction<? super T> shortFunction)
     {
         return Bags.immutable.withAll(this);
     }
+
+    @Override
+    public MutableSet<T> selectUnique()
+    {
+        MutableBag<T> selected = this.selectByOccurrences(occurrences -> occurrences == 1);
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/SynchronizedBag.java]: Same comment as primitive side.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/UnmodifiableBag.java [hunk 26]
```diff
@@ -426,4 +427,12 @@ protected Object readResolve()
             return this.mutableBag.asUnmodifiable();
         }
     }
+
+    @Override
+    public MutableSet<T> selectUnique()
+    {
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/UnmodifiableBag.java]: Same comment as primitive side.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/primitive/BooleanHashBag.java [hunk 27]
```diff
@@ -512,6 +513,14 @@ public MutableBooleanBag selectByOccurrences(IntPredicate predicate)
         return result;
     }
 
+    @Override
+    public MutableBooleanSet selectUnique()
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/mutable/primitive/BooleanHashBag.java]: same comment as primitive side.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/sorted/immutable/AbstractImmutableSortedBag.java [hunk 28]
```diff
@@ -461,4 +462,12 @@ public void reverseForEachWithIndex(ObjectIntProcedure<? super T> procedure)
     {
         throw new UnsupportedOperationException(this.getClass().getSimpleName() + ".asReversed() not implemented yet");
     }
+
+    @Override
+    public ImmutableSortedSet<T> selectUnique()
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/sorted/immutable/AbstractImmutableSortedBag.java]: I would suggest to move this implementation to `ImmutableSortedBagImpl` and have implementation using `forEachWithOccurrences` similar to other comments.
The `SortedSet` should inherit the comparator from ImmutableSortedBagImpl.

For ImmutableEmptySortedBag, simply return an empty ImmutableSortedSet which also inherits the comparator.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/sorted/mutable/AbstractMutableSortedBag.java [hunk 29]
```diff
@@ -352,4 +354,12 @@ public int detectLastIndex(Predicate<? super T> predicate)
         }
         throw new UnsupportedOperationException(this.getClass().getSimpleName() + ".asParallel() not implemented yet");
     }
+
+    @Override
+    public MutableSortedSet<T> selectUnique()
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/sorted/mutable/AbstractMutableSortedBag.java]:
```
MutableSortedSet<T> result = TreeBag.newBag(this.comparator());
this.forEachWithOccurrences((each, occurrences) ->
{
    if(occurrences == 1)
    {
        result.add(each);
    }
}
return result;
```

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/sorted/mutable/SynchronizedSortedBag.java [hunk 30]
```diff
@@ -716,4 +717,12 @@ public int detectLastIndex(Predicate<? super T> predicate)
             return this.getDelegate().getLastOptional();
         }
     }
+
+    @Override
+    public MutableSortedSet<T> selectUnique()
+    {
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/sorted/mutable/SynchronizedSortedBag.java]: Same comment as primitive side. Please wrap this around a synchronized block and delegate it.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/sorted/mutable/UnmodifiableSortedBag.java [hunk 31]
```diff
@@ -522,4 +523,12 @@ protected Object writeReplace()
     {
         return new UnmodifiableCollectionSerializationProxy<>(this.getSortedBag());
     }
+
+    @Override
+    public MutableSortedSet<T> selectUnique()
+    {
```
Comment (nikhilnanivadekar) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/sorted/mutable/UnmodifiableSortedBag.java]: Same comment as primitive side.

File: unit-tests/src/test/java/org/eclipse/collections/impl/bag/mutable/MultiReaderHashBagTest.java [hunk 32]
```diff
@@ -257,6 +257,15 @@ public void selectDuplicates()
                 bag.selectDuplicates());
     }
 
+    @Test
+    public void selectUnique()
+    {
+        MutableBag<Integer> bag = MultiReaderHashBag.newBagWith(0, 1, 1, 2, 2, 2, 3);
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/bag/mutable/MultiReaderHashBagTest.java]: 👍

File: unit-tests/src/test/java/org/eclipse/collections/impl/bag/mutable/MutableBagTestCase.java [hunk 33]
```diff
@@ -439,6 +439,13 @@ public void selectDuplicates()
         Assert.assertEquals(Iterables.iBag(1, 1, 1, 1, 2, 2, 2, 3, 3), integers.selectDuplicates());
     }
 
+    @Test
+    public void selectUnique()
+    {
+        MutableBagIterable<Integer> integers = this.newWith(0, 1, 1, 1, 1, 2, 2, 2, 3, 3, 4, 5);
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/bag/mutable/MutableBagTestCase.java]: 👍

File: unit-tests/src/test/java/org/eclipse/collections/impl/bag/mutable/SynchronizedBagTest.java [hunk 34]
```diff
@@ -146,6 +147,14 @@ public void selectDuplicates()
                 this.newWith(0, 1, 1, 1, 1, 2, 2, 2, 3, 3, 4, 5).selectDuplicates());
     }
 
+    @Test
+    public void selectUnique()
+    {
+        Assert.assertEquals(
+                iSet(0, 4, 5),
+                this.newWith(0, 1, 1, 1, 1, 2, 2, 2, 3, 3, 4, 5).selectUnique());
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/bag/mutable/SynchronizedBagTest.java]: 👍

File: unit-tests/src/test/java/org/eclipse/collections/impl/bag/mutable/UnmodifiableBagTest.java [hunk 35]
```diff
@@ -167,6 +167,14 @@ public void selectDuplicates()
                 UnmodifiableBag.of(HashBag.newBagWith(0, 1, 1, 1, 1, 2, 2, 2, 3, 3, 4, 5)).selectDuplicates());
     }
 
+    @Test
+    public void selectUnique()
+    {
+        Assert.assertEquals(
+              Sets.mutable.with(0, 4, 5),
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/bag/mutable/UnmodifiableBagTest.java]: 👍

File: unit-tests/src/test/java/org/eclipse/collections/impl/bag/sorted/mutable/AbstractMutableSortedBagTestCase.java [hunk 36]
```diff
@@ -1161,6 +1162,18 @@ public void selectDuplicates()
...
+    @Override
+    @Test
+    public void selectUnique()
+    {
+        super.selectUnique();
+
+        MutableSortedBag<Integer> integers = this.newWith(Collections.reverseOrder(), 5, 4, 3, 3, 2, 2, 2, 1, 1, 1, 1, 0);
+        Assert.assertEquals(
+                SortedSets.mutable.of(Collections.reverseOrder(), 5, 4, 0),
+                integers.selectUnique());
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/bag/sorted/mutable/AbstractMutableSortedBagTestCase.java]: Please add assertion for comparator.

**Comment by nikhilnanivadekar** [comment]
@loradd thank you very much for the PR. I have left a few comments to make the implementation more efficient.
Also, can you please squash all your commits as per the contribution guidelines: https://github.com/eclipse/eclipse-collections/blob/master/CONTRIBUTING.md#pull-requests

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/AbstractBag.java (hunk 20 above)
Comment (loradd) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/bag/AbstractBag.java]: This was an unwanted change and it has now been reverted. I was trying out a checkstyle plugin in Eclipse, sorry :)

**Comment by nikhilnanivadekar** [comment]
@loradd can you please rebase on to master? 
Sorry! It is difficult to review changes pertaining to your commit and separate them from changes already merged. You can refer to this wiki for more details: https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request

**Comment by loradd** [comment]
Done! Sorry for the inconvenience.

File: eclipse-collections-code-generator/src/main/resources/impl/bag/mutable/synchronizedPrimitiveBag.stg [hunk 37]
```diff
@@ -266,7 +268,15 @@ public class Synchronized<name>Bag
...
+    {
+       synchronized (this.getLock())
+       {
+            return this.getMutable<name>Bag().selectUnique();
+       }
+    }
 }
 
 >>
-
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/impl/bag/mutable/synchronizedPrimitiveBag.stg]: unnecessary formatting change. Can you please revert.

File: eclipse-collections-code-generator/src/main/resources/impl/bag/mutable/unmodifiablePrimitiveBag.stg [hunk 38]
```diff
@@ -177,7 +179,12 @@ public class Unmodifiable<name>Bag
...
+
+    @Override
+    public Mutable<name>Set selectUnique()
+    {
+        return this.getMutable<name>Bag().selectUnique();
+    }
 }
 
 >>
-
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/impl/bag/mutable/unmodifiablePrimitiveBag.stg]: Unnecessary formatting change. Can you please revert

File: eclipse-collections-code-generator/src/main/resources/test/bag/mutable/primitiveHashBagTest.stg [hunk 39]
```diff
@@ -82,6 +85,18 @@ public class <name>HashBagTest extends AbstractMutable<name>BagTestCase
         <wrapperName> sum = hashBag.injectInto(<wrapperName>.valueOf(<(literal.(type))("4")>), (<wrapperName> result, <type> value) -> <wrapperName>.valueOf((<type>) (result + value)));
         Assert.assertEquals(<wrapperName>.valueOf(<(literal.(type))("17")>), sum);
     }
+
+    @Override
+    @Test
+    public void selectUnique()
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/test/bag/mutable/primitiveHashBagTest.stg]: this override can probably be dropped since functionality is already tested by `AbstractMutable<name>BagTestCase`. Or may be I am missing something, please correct me if I am wrong.

File: eclipse-collections-code-generator/src/main/resources/test/bag/mutable/synchronizedPrimitiveBagTest.stg [hunk 40]
```diff
@@ -51,7 +53,18 @@ public class Synchronized<name>BagTest extends AbstractMutable<name>BagTestCase
         Assert.assertSame(bag, bag.asSynchronized());
         Assert.assertEquals(bag, bag.asSynchronized());
     }
+
+    @Override
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/test/bag/mutable/synchronizedPrimitiveBagTest.stg]: this override can be removed as well.

File: eclipse-collections-code-generator/src/main/resources/test/bag/mutable/unmodifiablePrimitiveBagTest.stg [hunk 41]
```diff
@@ -261,6 +263,18 @@ public class Unmodifiable<name>BagTest extends AbstractMutable<name>BagTestCase
     {
         // Not applicable for Unmodifiable*
     }
+
+    @Override
+    @Test
+    public void selectUnique()
```
Comment (nikhilnanivadekar) [inline eclipse-collections-code-generator/src/main/resources/test/bag/mutable/unmodifiablePrimitiveBagTest.stg]: this override can be removed as well.

File: unit-tests/src/test/java/org/eclipse/collections/impl/bag/immutable/primitive/ImmutableBooleanSingletonBagTest.java [hunk 42]
```diff
@@ -11,11 +11,12 @@
 package org.eclipse.collections.impl.bag.immutable.primitive;
 
 import org.eclipse.collections.api.bag.primitive.ImmutableBooleanBag;
+import org.eclipse.collections.api.set.primitive.ImmutableBooleanSet;
 import org.eclipse.collections.impl.factory.primitive.BooleanBags;
+import org.eclipse.collections.impl.factory.primitive.BooleanSets;
 import org.eclipse.collections.impl.test.Verify;
 import org.junit.Assert;
 import org.junit.Test;
-
```
Comment (nikhilnanivadekar) [inline unit-tests/src/test/java/org/eclipse/collections/impl/bag/immutable/primitive/ImmutableBooleanSingletonBagTest.java]: Please add back the removed blank line.

File: eclipse-collections-code-generator/src/main/resources/test/bag/mutable/primitiveHashBagTest.stg (hunk 39 above)
Comment (loradd) [reply eclipse-collections-code-generator/src/main/resources/test/bag/mutable/primitiveHashBagTest.stg]: I agree, the one in `AbstractMutable<name>BagTestCase` already covers the tested functionality.

**Comment by loradd** [comment]
@nikhilnanivadekar the "Java 11 regression" build job is failing, can you help me with that?

**Comment by nikhilnanivadekar** [comment]
You can ignore that build failure for now.

**Comment by nikhilnanivadekar** [comment]
Thanks for your contribution!

**Comment by loradd** [comment]
Glad to contribute, thank you for the support!
